code/datums/components/scope.dm

```diff
@@ -0,0 +1,274 @@
+		var/obj/item/parent_item = parent
+		var/datum/action/scope = locate(item_action_type) in parent_item.actions
+		parent_item.actions -= (scope)
+	UnregisterSignal(parent, list(
+		COMSIG_MOVABLE_MOVED,
+		SIGNAL_ADDTRAIT(TRAIT_WIELDED),
+		SIGNAL_REMOVETRAIT(TRAIT_WIELDED),
+		COMSIG_GUN_TRY_FIRE,
+		COMSIG_PARENT_EXAMINE,
+	))
+
+/datum/component/scope/process()
+	var/mob/user_mob = tracker.owner
+	var/client/user_client = user_mob.client
+	if(!user_client)
+		stop_zooming(user_mob)
+		return
+	tracker.calculate_params()
+	animate(user_client, world.tick_lag, pixel_x = tracker.given_x, pixel_y = tracker.given_y)
+
+/datum/component/scope/proc/on_move(atom/movable/source, atom/oldloc, dir, forced)
+	SIGNAL_HANDLER
+
+	if(!tracker)
+		return
+	stop_zooming(tracker.owner)
+
+/datum/component/scope/proc/on_action_trigger(datum/action/source)
+	SIGNAL_HANDLER
+	var/obj/item/item = source.target
+	var/mob/living/user = item.loc
+	if(is_internal_organ(item))
+		var/obj/item/organ/internal/O = item
+		user = O.owner
+	if(tracker)
+		stop_zooming(user)
+	else
+		INVOKE_ASYNC(src, PROC_REF(zoom), user)
+
+/datum/component/scope/proc/on_wielded(obj/item/source, trait)
+	SIGNAL_HANDLER
+	var/mob/living/user = source.loc
+	INVOKE_ASYNC(src, PROC_REF(zoom), user)
+
+/datum/component/scope/proc/on_unwielded(obj/item/source, trait)
+	SIGNAL_HANDLER
+	var/mob/living/user = source.loc
+	stop_zooming(user)
+
+/datum/component/scope/proc/on_gun_fire(obj/item/gun/source, mob/living/user, atom/target, flag, params)
+	SIGNAL_HANDLER
+	if(!tracker?.given_turf || target == get_target(tracker.given_turf))
+		return NONE
+	INVOKE_ASYNC(source, TYPE_PROC_REF(/obj/item, afterattack), get_target(tracker.given_turf), user)
+	return COMPONENT_CANCEL_GUN_FIRE
+
+/datum/component/scope/proc/on_examine(datum/source, mob/user, list/examine_list)
+	SIGNAL_HANDLER
+
+	var/scope = istype(parent, /obj/item/gun) ? "scope in" : "zoom out"
+	switch(zoom_method)
+		if(ZOOM_METHOD_WIELD)
+			examine_list += "<span class='notice'>You can [scope] by wielding it with both hands.</span>"
+
+/**
+ * We find and return the best target to hit on a given turf.
+ *
+ * Arguments:
+ * * target_turf: The turf we are looking for targets on.
+*/
+/datum/component/scope/proc/get_target(turf/target_turf)
+	var/list/object_targets = list()
+	var/list/non_dense_targets = list()
+	for(var/atom/movable/possible_target in target_turf)
+		if(possible_target.layer <= PROJECTILE_HIT_THRESHHOLD_LAYER)
+			continue
+		if(possible_target.invisibility > tracker.owner.see_invisible)
+			continue
+		if(!possible_target.mouse_opacity)
+			continue
+		if(iseffect(possible_target))
+			continue
+		if(ismob(possible_target))
+			if(possible_target == tracker.owner)
+				continue
+			return possible_target
+		if(!possible_target.density)
+			non_dense_targets += possible_target
+			continue
+		object_targets += possible_target
+	for(var/obj/important_object as anything in object_targets)
+		return important_object
```

Comment:
```suggestion
		return possible_target
```
Why not return it here? You return the first found value anyways. Unless you wanted to do a `return pick(object_targets)` after this for loop